client-go/src/main/client/session.go

```diff
@@ -0,0 +1,698 @@
+	status, err := s.client.TestInsertStringRecords(context.Background(), &request)
+	if err != nil {
+		Log.WithError(err).Error("insert multi string records failed")
+	} else {
+		Log.WithFields(logrus.Fields{
+			"dv":   s.dv,
+			"code": status.Code,
+		}).Info("insert multi string records success")
+	}
+}
+
+/*
+ *insert one row of record into database, if you want improve your performance, please use insertTablet method
+ *
+ *params
+ *deviceId: string, time series path for device
+ *measurements: []string, sensor names
+ *dataTypes: []int32, list of dataType, indicate the data type for each sensor
+ *values: []interface{}, values to be inserted, for each sensor
+ *timestamp: int64, indicate the timestamp of the row of data
+ *
+ *return
+ *error: correctness of operation
+ */
+func (s *Session) InsertRecord(deviceId string, measurements []string, dataTypes []int16, values []interface{},
+	timestamp int64) {
+	request := s.genInsertRecordReq(deviceId, measurements, dataTypes, values, timestamp)
+	status, err := s.client.InsertRecord(context.Background(), request)
+	if err != nil {
+		Log.WithError(err).Error("insert one record failed")
+	} else {
+		Log.WithFields(logrus.Fields{
+			"dv":   deviceId,
+			"code": status.Code,
+		}).Info("insert one record success")
+	}
+}
+
+func (s *Session) TestInsertRecord(deviceId string, measurements []string, dataTypes []int16, values []interface{},
+	timestamp int64) {
+	request := s.genInsertRecordReq(deviceId, measurements, dataTypes, values, timestamp)
+	status, err := s.client.TestInsertRecord(context.Background(), request)
+	if err != nil {
+		Log.WithError(err).Error("insert one record failed")
+	} else {
+		Log.WithFields(logrus.Fields{
+			"dv":   deviceId,
+			"code": status.Code,
+		}).Info("insert one record success")
+	}
+}
+
+func (s *Session) genInsertRecordReq(deviceId string, measurements []string, dataTypes []int16, values []interface{},
+	timestamp int64) *rpc.TSInsertRecordReq {
+	request := rpc.TSInsertRecordReq{SessionId: s.sessionId, DeviceId: deviceId, Measurements: measurements,
+		Timestamp: timestamp}
+	request.Values = valuesToBytes(dataTypes, values)
+	return &request
+}
+
+/*
+ *insert multiple rows of data, records are independent to each other, in other words, there's no relationship
+ *between those records
+ *
+ *params
+ *deviceIds: []string, time series paths for device
+ *measurements: [][]string, each element of outer list indicates measurements of a device
+ *dataTypes: [][]int32, each element of outer list indicates sensor data types of a device
+ *values: [][]interface{}, values to be inserted, for each device
+ *timestamps: []int64, timestamps for records
+ *
+ *return
+ *error: correctness of operation
+ */
+func (s *Session) InsertRecords(deviceIds []string, measurements [][]string, dataTypes [][]int16, values [][]interface{},
+	timestamps []int64) {
+	s.dv = strings.Replace(strings.Trim(fmt.Sprint(deviceIds), "[]"), " ", ",", -1)
```

Comment:
for every inset operation, this may time-consume, If you want to print the devices,  I think the `deviceIds[]` is just enough.

Reply:
But the direct output deviceids [] is the address

Reply:
then you can do it if the log debug level is enabled, if the log level is info,  for every operation the code will be called once, it's unnecessary!